[nethost] Update nethost

lanyizi opened the pull request:
**Describe the pull request**

- #### What does your PR fix?
  Addressed the issue with static library. The previous version was pushed long ago, back then `__declspec(dllimport)` was unconditionally enabled on the `nethost.h` when compiling Windows platform, which hampers its use as static library: https://github.com/dotnet/runtime/blob/188427d7e18102c45fc6d0e20c135e226f215992/src/installer/corehost/cli/nethost/nethost.h#L12
  This PR updates the version of referenced .NET code base to .NET 6, which has a better support for static library: https://github.com/dotnet/runtime/blob/70ae3df4a6f3c92fb6b315afc405edd10ff38579/src/native/corehost/nethost/nethost.h#L15
  This PR also adds imported targets, allowing users to `find_package()` and then `target_link_library` directly, without having to manually set up include directories or compiler definitions

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  Triplet and CI baseline should be same as before, although personally I couldn't test all platforms.

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  I guess yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

lanyizi committed:
update nethost

lanyizi committed:
bump nethost version

FrankXie05 commented on ports/nethost/portfile.cmake:
```diff
@@ -53,13 +57,15 @@ vcpkg_configure_cmake(
 
 vcpkg_copy_pdbs()
 
+vcpkg_fixup_cmake_targets()
```
```suggestion
vcpkg_cmake_config_fixup()
```

FrankXie05 commented on ports/nethost/vcpkg.json:
```diff
@@ -1,7 +1,7 @@
-  "version-date": "2019-12-21",
-  "port-version": 2,
+  "version-date": "2022-05-30",
+  "port-version": 3,
```
```suggestion
```
Please bump the port version. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#update-the-version-field-in-the-control-file-of-any-modified-ports).
Also please run command `vcpkg x-add-version --all` then commit the changes.

FrankXie05 reviewed (changes requested):
Please note that these functions are about to be deprecated, please use the new method.
https://github.com/microsoft/vcpkg/pull/25004#pullrequestreview-989211355

lanyizi committed:
Add CMake exports in an unofficial- namespace
https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#add-cmake-exports-in-an-unofficial--namespace

lanyizi committed:
fix port version and deprecated functions

lanyizi commented on ports/nethost/portfile.cmake:
```diff
@@ -41,9 +40,14 @@ endif()
+set(NETHOST_SOURCE_PATH ${SOURCE_PATH}/src/native/corehost/nethost/)
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/nethost-config.cmake.in DESTINATION ${NETHOST_SOURCE_PATH})
+
 vcpkg_configure_cmake(
```
I changed the function, however when I'm running a test build I got this error:
`Unknown CMake command "vcpkg_cmake_configure".`
I'm very sure that I have updated vcpkg (its version is also shown in the following log), and... I have no idea why this happens...


<details>
  <summary>Error Log</summary>
```
Detecting compiler hash for triplet x86-windows-static...
The following packages will be built and installed:
    nethost[core]:x86-windows-static -> 2022-05-30
Restored 0 packages from C:\Users\lanyi\AppData\Local\vcpkg\archives in 1.361 ms. Use --debug to see more details.
Installing 1/1 nethost:x86-windows-static...
Building nethost[core]:x86-windows-static...
-- Using community triplet x86-windows-static. This triplet configuration is not guaranteed to succeed.
-- [COMMUNITY] Loading triplet configuration from: C:\Users\lanyi\source\repos\vcpkg\triplets\community\x86-windows-static.cmake
-- Using cached dotnet-runtime-v6.0.5.tar.gz.
-- Cleaning sources at C:/Users/lanyi/source/repos/vcpkg/buildtrees/nethost/src/v6.0.5-4fb79eaecc.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/Users/lanyi/source/repos/vcpkg/downloads/dotnet-runtime-v6.0.5.tar.gz
-- Applying patch 0001-nethost-cmakelists.patch
-- Using source at C:/Users/lanyi/source/repos/vcpkg/buildtrees/nethost/src/v6.0.5-4fb79eaecc.clean
CMake Error at ports/nethost/portfile.cmake:46 (vcpkg_cmake_configure):
  Unknown CMake command "vcpkg_cmake_configure".
Call Stack (most recent call first):
  scripts/ports.cmake:146 (include)


error: building nethost:x86-windows-static failed with: BUILD_FAILED
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
    https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+nethost
You can submit a new issue at:
    https://github.com/microsoft/vcpkg/issues/new?template=report-package-build-failure.md&title=[nethost]+Build+error
Include '[nethost] Build error' in your bug report title, the following version information in your bug description, and attach any relevant failure logs from above.
    vcpkg-tool version: 2022-05-05-67e17c1782801cf481be9ac0b3765dff3e4bdeb8
    vcpkg-scripts version: 51f26704e 2022-05-31 (2 minutes ago)
```
</details>

FrankXie05 replied:
New functions need dependencies, add dependencies in `vcpkg.json`:
````
  "dependencies": [
    {
      "name": "vcpkg-cmake",
      "host": true
    },
    {
      "name": "vcpkg-cmake-config",
      "host": true
    }
  ],
  ````

lanyizi commented:
I have updated the deprecated functions, for example replacing `vcpkg_configure_cmake` with `vcpkg_cmake_configure`, however now the build fails, saying "Unknown CMake command: vcpkg_cmake_configure"...
https://dev.azure.com/vcpkg/public/_build/results?buildId=72951&view=logs&j=878666d5-db33-5b27-9e7d-b0c7ee352005&t=3980505e-92f5-53ef-75a8-4f2e70b97516&l=6015

FrankXie05 commented:
@lanyizi  New functions need dependencies, add dependencies in `vcpkg.json`:
````
  "dependencies": [
    {
      "name": "vcpkg-cmake",
      "host": true
    },
    {
      "name": "vcpkg-cmake-config",
      "host": true
    }
  ],
  ````

FrankXie05 commented on ports/nethost/vcpkg.json:
```diff
@@ -1,7 +1,6 @@
+  "version-date": "2022-05-30",
   "description": "Provides a set of APIs that can be used to host .NET Core (dotnet)",
   "homepage": "https://github.com/dotnet/runtime/tree/master/src/installer/corehost/cli/nethost",
   "supports": "!uwp"
```
```suggestion
  "supports": "!uwp"
  "dependencies": [
    {
      "name": "vcpkg-cmake",
      "host": true
    },
    {
      "name": "vcpkg-cmake-config",
      "host": true
    }
  ]
```
Remember to run the command `.\vcpkg.exe format-manifest  .\ports\nethost\vcpkg.json to format the `vcpkg.json` file after adding these.

lanyizi replied:
Thanks for the info!

FrankXie05 commented on ports/nethost/vcpkg.json:
```diff
@@ -1,7 +1,6 @@
   "name": "nethost",
-  "version-date": "2019-12-21",
-  "port-version": 2,
+  "version-date": "2022-05-30",
```
```suggestion
  "version": "6.0.5",
```
I think that since the upstream has updated the specific version number, `version-data` can be discarded. This will make subsequent version maintenance and updates more clear.

lanyizi committed:
change nethost version date to version

FrankXie05 commented on ports/nethost/usage:
```diff
@@ -1,6 +1,5 @@
-    find_library(NETHOST_LIBRARY NAMES libnethost nethost)
-    target_include_directories(main PRIVATE ${NETHOST_INCLUDE_DIRS})
-    target_link_libraries(main PRIVATE ${NETHOST_LIBRARY})
+    find_package(unofficial-nethost CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE unofficial::nethost::nethost) # DLL
+    target_link_libraries(main PRIVATE unofficial::nethost::libnethost) # Static
```
```
    find_package(unofficial-nethost CONFIG REQUIRED)
    target_link_libraries(main PRIVATE unofficial::nethost::nethost) # DLL
    target_link_libraries(main PRIVATE unofficial::nethost::libnethost) # Static
```
Is it necessary to separate the linking of dynamic and static libraries?

FrankXie05 replied:
And this usage is not available.
````
CMake Error at /Users/vcpkg/Frank/vcpkg/scripts/buildsystems/vcpkg.cmake:573 (_add_executable):
  Target "test" links to target "unofficial::nethost::nethost" but the target
  was not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  CMakeLists.txt:5 (add_executable)
 ````
Example CMakeLists.txt:
````
  1 cmake_minimum_required(VERSION 3.8)
  2
  3 project(test)
  4
  5 add_executable(test "test.cpp")
  6
  7 find_package(unofficial-nethost CONFIG REQUIRED)
  8 target_link_libraries(test PRIVATE unofficial::nethost::nethost) # DLL
  9 target_link_libraries(test PRIVATE unofficial::nethost::libnethost) # Static
````

lanyizi replied:
`unofficial::nethost::nethost` was meant to be used for non-static triplet
So if the current triplet is something like `x86-windows-static`, then only `unofficial::nethost::libnethost` is available.


This is also because upstream decided to name the two targets differently, they build the same source twice (both as static and shared library) at the same time:
https://github.com/dotnet/runtime/blob/70ae3df4a6f3c92fb6b315afc405edd10ff38579/src/native/corehost/nethost/CMakeLists.txt#L22
https://github.com/dotnet/runtime/blob/70ae3df4a6f3c92fb6b315afc405edd10ff38579/src/native/corehost/nethost/CMakeLists.txt#L23

If this appears too weird I can edit the patch file so both target have same name. Should I do that?

FrankXie05 commented:
I found that for upstream they don't want nethost to export both dynamic and static libraries. https://github.com/dotnet/runtime/blob/55e2378d86841ec766ee21d5e504d7724c39b53b/src/native/corehost/nethost/nethost.h#L13. 
There is nothing wrong with using the macro `NETHOST_USE_AS_STATIC` to differentiate, but can we add an option 
like `-DBUILD_SHARED` or `-DBUILD_STATIC` to differentiate the usage of the exported library?(Maybe this way of yours is better, I'm not sure. I am testing your usage) 🤔

dg0yt commented:
> can we add an option
> like `-DBUILD_SHARED` or `-DBUILD_STATIC` to differentiate the usage of the exported library?

Don't invent a new option. Just leverage `BUILD_SHARED_LIBS`.
https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html

lanyizi commented:
> Don't invent a new option. Just leverage `BUILD_SHARED_LIBS`. https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html

I feel `BUILD_SHARED_LIBS` won't have any effect when building the nethost, because upstream explicitly defined shared / dynamic targets which are being built at the same time...
https://github.com/dotnet/runtime/blob/v6.0.5/src/native/corehost/lib.cmake
https://github.com/dotnet/runtime/blob/v6.0.5/src/native/corehost/lib_static.cmake

dg0yt commented:
> I feel `BUILD_SHARED_LIBS` won't have any effect when building the nethost, because upstream explicitly defined shared / dynamic targets which are being built at the same time...

Yes, this CMake anti-pattern is known from other ports. But vcpkg wants only one variant per triplet.
- You can add patches which use `BUILD_SHARED_LIBS` for `if/else` in the CMake build system. 
  `BUILD_SHARED_LIBS` is already passed by vcpkg as needed by the triplet.
- You can give a `usage` example which makes use of generator expressions to select the library name.
  Example: Port `zstd`, https://github.com/microsoft/vcpkg/blob/d953973cfbfeaf8022f5219fbabb0bdf9defb706/ports/zstd/usage#L1-L4

lanyizi commented:
> * You can give a `usage` example which makes use of generator expressions to select the library name.
>   Example: Port `zstd`, https://github.com/microsoft/vcpkg/blob/d953973cfbfeaf8022f5219fbabb0bdf9defb706/ports/zstd/usage#L1-L4

I feel this is easier to do for me (

lanyizi committed:
update usage

FrankXie05 commented:
cc @JackBoosY to review again.

JackBoosY commented on ports/nethost/0001-nethost-cmakelists.patch:
```diff
@@ -1,58 +1,78 @@
++
  include(../lib_static.cmake)
-+endif()
++target_compile_definitions(libnethost PUBLIC NETHOST_USE_AS_STATIC)
```
Only support static build?

JackBoosY commented on ports/nethost/portfile.cmake:
```diff
@@ -41,9 +40,13 @@ endif()
-vcpkg_configure_cmake(
-    SOURCE_PATH ${SOURCE_PATH}/src/installer/corehost/cli/nethost
-    PREFER_NINJA
+set(NETHOST_SOURCE_PATH ${SOURCE_PATH}/src/native/corehost/nethost/)
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/unofficial-nethost-config.cmake.in DESTINATION ${NETHOST_SOURCE_PATH})
+
+vcpkg_cmake_configure(
+    SOURCE_PATH ${NETHOST_SOURCE_PATH}
```
```suggestion
set(NETHOST_SOURCE_PATH "${SOURCE_PATH}/src/native/corehost/nethost/")
file(COPY "${CMAKE_CURRENT_LIST_DIR}/unofficial-nethost-config.cmake.in" DESTINATION "${NETHOST_SOURCE_PATH}")

vcpkg_cmake_configure(
    SOURCE_PATH "${NETHOST_SOURCE_PATH}"
```

JackBoosY commented on ports/nethost/unofficial-nethost-config.cmake.in:
```diff
@@ -0,0 +1,2 @@
+@PACKAGE_INIT@
+include("${CMAKE_CURRENT_LIST_DIR}/unofficial-nethost-targets.cmake")
```
Why not directly export the targets as `unofficial-nethost-config` instead?

JackBoosY commented on ports/nethost/usage:
```diff
@@ -1,6 +1,6 @@
-    find_library(NETHOST_LIBRARY NAMES libnethost nethost)
-    target_include_directories(main PRIVATE ${NETHOST_INCLUDE_DIRS})
-    target_link_libraries(main PRIVATE ${NETHOST_LIBRARY})
+    find_package(unofficial-nethost CONFIG REQUIRED)
```
Since the targets are exported, I think we don't need this usage file now.

lanyizi committed:
add quotes to paths and export targets directly

JackBoosY commented on ports/nethost/0001-nethost-cmakelists.patch:
```diff
@@ -1,58 +1,70 @@
++endif(BUILD_SHARED_LIBS)
++
++target_include_directories(${NETHOST_ACTIVE_TARGET} INTERFACE $<INSTALL_INTERFACE:include>)
++install(TARGETS ${NETHOST_ACTIVE_TARGET} EXPORT unofficial-nethost-targets
```
```suggestion
+install(TARGETS ${NETHOST_ACTIVE_TARGET} EXPORT unofficial-nethost-config

```

JackBoosY commented on ports/nethost/0001-nethost-cmakelists.patch:
```diff
@@ -1,58 +1,70 @@
++    RUNTIME DESTINATION bin
++)
++
++install(EXPORT unofficial-nethost-targets
```
```suggestion
+install(EXPORT unofficial-nethost-config
```

JackBoosY commented:
Never mind the requests, did you test the usage locally?

lanyizi commented:
Yes, I tested it with both `x86-windows` and `x86-windows-static` triplet
If needed I think I could also test x64-windows, but personally I can't test on other platforms (arm or linux)

lanyizi committed:
bump nethost version

ras0219-msft committed:
[nethost] Only build one flavor

ras0219-msft reviewed (changes requested):
Thanks for the PR!

I have reworked the patch to do a few additional things:
1. Always install the name "unofficial::nethost::nethost", to avoid users needing to write an awful generator expression. If a static build is requested, it's just a link to the `libnethost` target
2. Embed the library linkage choice into the header
3. I searched the source and found no uses of `CLI_CMAKE_PLATFORM_ARCH`, but I did see `CLR_CMAKE_TARGET_ARCH`, so I renamed that value.

Please run the following to merge those commits into this PR:

```
git fetch https://github.com/ras0219-msft/vcpkg dev/roschuma/nethost
git merge FETCH_HEAD
```

lanyizi commented:
Thanks for the rework! I have merged the new commit into PR

lanyizi committed:
fix build error

lanyizi commented:
@ras0219-msft Hi! I noticed there was some build errors, for example now the static build didn't `include(../lib.cmake)` anymore, however `lib.cmake` introduced some source files which are needed by the static build as well.
I edited the patch file so those source files are added to static target too.

lanyizi committed:
bump nethost version

ras0219-msft reviewed (approved):
Thanks!
